x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -631,31 +618,41 @@ public SourceProvider createSourceProvider(SourceFilter sourceFilter) {
                 enrichLookupService,
                 lookupFromIndexService,
                 inferenceService,
-                physicalOperationProviders,
-                contexts
+                physicalOperationProviders
             );
 
-            LOGGER.debug("Received physical plan:\n{}", plan);
+            LOGGER.debug("Received physical plan for {}:\n{}", context.description(), plan);
 
             var localPlan = PlannerUtils.localPlan(
                 plannerSettings,
                 context.flags(),
-                context.searchExecutionContexts(),
+                new ArrayList<>(context.searchExecutionContexts().collection()),
                 context.configuration(),
                 context.foldCtx(),
                 plan
             );
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Local plan for {}:\n{}", context.description(), localPlan);
+            }
             // the planner will also set the driver parallelism in LocalExecutionPlanner.LocalExecutionPlan (used down below)
             // it's doing this in the planning of EsQueryExec (the source of the data)
             // see also EsPhysicalOperationProviders.sourcePhysicalOperation
-            LocalExecutionPlanner.LocalExecutionPlan localExecutionPlan = planner.plan(context.description(), context.foldCtx(), localPlan);
+            LocalExecutionPlanner.LocalExecutionPlan localExecutionPlan = planner.plan(
+                context.description(),
+                context.foldCtx(),
+                localPlan,
+                context.searchContexts().map(ComputeSearchContext::shardContext)
+            );
             if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug("Local execution plan:\n{}", localExecutionPlan.describe());
+                LOGGER.debug("Local execution plan for {}:\n{}", context.description(), localExecutionPlan.describe());
             }
             var drivers = localExecutionPlan.createDrivers(context.sessionId());
-            // After creating the drivers (and therefore, the operators), we can safely decrement the reference count since the operators
-            // will hold a reference to the contexts where relevant.
-            contexts.forEach(RefCounted::decRef);
+            // After creating the drivers (and therefore, the operators), we can safely decrement the reference count since the reader
+            // operators will hold a reference to the contexts where relevant. However, we should only do this for the data computations,
+            // because in other cases the drivers won't increment the reference count of the contexts (no readers).
```

Comment:
Hold on. This seems to imply that in the non-data computations, the drivers need the shard contexts but don't incRef them? I don't know if this is right. My instinct is that we should always decref here, because this piece of code doesn't need to hold on to the shard contexts anymore. If another piece of code still needs the shard contexts, it should incref them and take care of properly closing them once done with them.

Reply:
yeah, I think words are hard.

What he means is that the coordinating driver shouldn't inc the ref. It gets it's refs from the doc blocks it receives. If it doesn't contain any docs from that context, we don't need to leave it open. If it does, we leave it open. If the "data" driver - the "leaf" driver - is done with a doc and the doc block doesn't hold the counter, then we close the context.

By the same token, only the `LuceneSourceOperator`s actually need to hold the reference. Everything after that can use the reference passed to them in the DocBlock. Once a source no longer needs a reader *AND* there are no `DocBlock`s pointing at that read alive, we can close the reader. We'll never need it again.

Modulo TopN which has to hold a reference to each doc block it might read.

Reply:
Ah! That makes sense. I think.

It'd be lovely to describe this in the comment @GalLalouche , because looking from outside it just looks like the reduce driver will have contexts that may or may not be closed already, depending on what the data drivers do, and it's not intuitive that the contexts being already closed is correct because the reduce driver will not require them (no doc blocks passed to it, as I understand).